### ISSUE TITLE
libcurl 7.72.0: CURLE_SEND_ERROR from simple GET request on SLES 12 SP 4

### ISSUE THREAD

**Issue by ndchoate** [curl/curl#6001]
Hello, please bear with me if this issue does not follow the proper etiquette, this is my first time posting one. Please let me know if you need more details. Thanks in advance!

### I did this
We recently upgraded libcurl from version 7.71.1 to 7.72.0. After doing so, our integration tests started failing intermittently on SLES 12 SP 4 due to connections dying on simple GET requests that utilize the easy API. Here is the debug output from libcurl when the issue is encountered:

```
DEBUG: Re-using existing connection! (#4) with host <local-SuSE-machine>
DEBUG: Connected to <local-SuSE-machine> () port 4567 (#4)
DEBUG: Send failure: Broken pipe
DEBUG: Failed sending HTTP request
DEBUG: Connection died, retrying a fresh connect(retry count: 5)
DEBUG: Closing connection 4
DEBUG: Issue another request to this URL: 'http://<local-SuSE-machine>:4567/<path>'

...

DEBUG: Found bundle for host <local-SuSE-machine>: 0xb988550 [serially]
DEBUG: Re-using existing connection! (#5) with host <local-SuSE-machine>
DEBUG: Connected to <local-SuSE-machine> () port 4567 (#5)
DEBUG: Send failure: Broken pipe
DEBUG: Failed sending HTTP request
DEBUG: Connection died, tried 5 times before giving up
DEBUG: Connection #5 to host <local-SuSE-machine> left intact
WARNING: (From our client, which utilizes libcurl): http://<local-SuSE-machine>:4567/<path> GET failed: 55 CURLE_SEND_ERROR: Failed sending network data.
DEBUG: (From our client, which utilizes libcurl): 55 CURLE_SEND_ERROR: Failed sending network data.
```

The requests are made to a Jetty-based server that is on the same machine as the client. This seems to only happen on SLES 12 SP 4. Our integration tests are passing for RHEL 7.6 and Windows Server 2016. AFAIK, there have been no changes to the configuration on the SLES machines.

I'm not sure if this will help, but below is some sample code of the simple GET request we execute.

```
void exampleSimpleDownload()
{
	CURL *pCurl = curl_easy_init();
	char *url = "http://<local-SuSE-machine>:4567/<path>"

	curl_easy_setopt(pCurl, CURLOPT_HTTP_VERSION, (long)CURL_HTTP_VERSION_1_1);
	curl_easy_setopt(pCurl, CURLOPT_HEADER, FALSE);
	curl_easy_setopt(pCurl, CURLOPT_NOSIGNAL, TRUE);
	curl_easy_setopt(pCurl, CURLOPT_ERRORBUFFER, errorMessage);
	curl_easy_setopt(pCurl, CURLOPT_BUFFERSIZE, 64 * 1024);
	curl_easy_setopt(pCurl, CURLOPT_TCP_NODELAY, TRUE);
	curl_easy_setopt(pCurl, CURLOPT_USERAGENT, userAgent);
	curl_easy_setopt(pCurl, CURLOPT_CONNECTTIMEOUT, 30);
	curl_easy_setopt(pCurl, CURLOPT_NOPROGRESS, TRUE);

	curl_easy_setopt(pCurl, CURLOPT_URL, url);

	curl_easy_setopt(pCurl, CURLOPT_HTTPGET, 1L);
	curl_easy_setopt(pCurl, CURLOPT_WRITEFUNCTION, writeCallback);
	curl_easy_setopt(pCurl, CURLOPT_WRITEDATA, writeData);
	curl_easy_setopt(pCurl, CURLOPT_HEADERFUNCTION, headerCallback);
	curl_easy_setopt(pCurl, CURLOPT_WRITEHEADER, writeHeader);

	curl_easy_perform(pCurl);
}
```

For testing purposes, we changed CONN_MAX_RETRIES in transfer.c to 30. This seems to have "fixed" the issue, but we have seen retry attempts ranging anywhere from 3 to 25.

### I expected the following
Successful execution of the GET requests with little to no retry attempts.

### curl/libcurl version

libcurl 7.72.0

### operating system
Linux vcl6005 4.12.14-94.41-default #1 SMP Wed Oct 31 12:25:04 UTC 2018 (3090901) x86_64 x86_64 x86_64 GNU/Linux

SuSE Linux Enterprise Server 12 Service Pack 4

**Comment by bagder** [comment]
Without a way for us to repro, you're left to debug this yourself. It seems the "Send failure: Broken pipe" is a key here. Why does the send fail? Does the server log give any hints?

You could possibly try to [bisect](https://github.com/curl/curl/wiki/how-to-git-bisect) to the single commit that introduced this problem for you, as it might help us understand what's going on.

**Comment by ndchoate** [comment]
Thanks for the response! Unfortunately, the server logs don't seem to give any idea as to why the send failed. The most that we have to go off of is a packet capture that shows a TCP RST coming from the server right before curl encounters the error.

It's worth noting now that we've been able to replicate this on Windows Server 2016 as well. However, it's only reproducible on one of our Windows Server machines. Another Windows Server machine (a VM on the same server) has no issues at all. It's quite puzzling because this makes it seem the issue lies with some configuration on the machines, but the problem only arises while using libcurl 7.72.0.

I think we're going to try the bisect route. Hopefully that will shed some light on the root cause of these issues.

**Comment by bagder** [comment]
> shows a TCP RST coming from the server right before curl encounters the error.

That sounds like a server issue? If no data has been transmitted, I take it curl retries the request over a new connection? Does that connection then also get a RST back?

**Comment by ndchoate** [comment]
Yes, it's certainly starting to seem like a server issue. Since the issue only started happening from the upgrade from 7.71.1 to 7.72.0, we thought it may be a client-side issue. We were thinking that perhaps there was some issue with the way we are utilizing curl, and were curious if there were any known issues or recent changes surrounding this scenario.

Based on the packet capture, it seems there's multiple TCP RSTs. Most of the time, curl is able to recover, but it seems that the problem occurs when a final TCP RST is sent and curl is not able to retry the request within 5 attempts.

I'm starting to think that the curl upgrade is just showing new symptoms of some problem server-side that we may have been experiencing for awhile. Nevertheless, we're now trying the git bisect route as well as debugging the server side. 

If you prefer, we can close the issue and I can update later on if we find anything interesting from a curl perspective. Thanks again, I very much appreciate the response!

**Comment by bagder** [comment]
@ndchoate, let's keep it open for another week just in case you find something additional to add that changes the story. If not, we can close it next week. Thanks for the added details!

**Comment by bagder** [comment]
A week passed and we have no additional info. Closing.

**Comment by ndchoate** [comment]
Sorry for the late update. I had only found the root cause a couple of days ago and I wanted to do some extensive testing before responding.

This was in fact a server-side issue, specifically with Jetty. The root cause involved uploads using a chunked transfer encoding. In certain situations, we would send a response for a chunked upload with a multipart form-data body before Curl had the chance to send the terminating chunk. This put Jetty in a weird state for the next request, causing it to send a TCP RST to Curl and making it unresponsive for some time (in this case, for the duration of time that Curl was attempting to reissue the failed GET request). A similar problem is described in [this issue](https://github.com/eclipse/jetty.project/issues/4117) on Jetty's GitHub.

The problem was that in our Jetty application, we consumed the input stream only up to the final character of the last multipart form-data boundary. To make sure that Jetty had the chance to consume and parse the terminating chunk, we had to do one final read from the input stream and receive a -1 from Jetty (as described in the link above).

I don't think this points to any problems with Curl, but I thought it would be good to update the issue for other people's reference. Thanks again for the help!
